fix(pageserver): abort on duplicate layers, before doing damage

problame opened the pull request:
fixes https://github.com/neondatabase/neon/issues/7790

# Background

From the time before always-authoritative `index_part.json`, we had to handle duplicate layers. See  the RFC for an illustration of how duplicate layers could happen: https://github.com/neondatabase/neon/blob/a8e6d259cb49d1bf156dfc2215b92c04d1e8a08f/docs/rfcs/027-crash-consistent-layer-map-through-index-part.md?plain=1#L41-L50 

As of #5198 , we should not be exposed to that problem anymore.

# Problem 1

We still have
1. [code in Pageserver](https://github.com/neondatabase/neon/blob/82960b2175211c0f666b91b5258c5e2253a245c7/pageserver/src/tenant/timeline.rs#L4502-L4521) than handles duplicate layers
2. [tests in the test suite](https://github.com/neondatabase/neon/blob/d9dcbffac37ccd3331ec9adcd12fd20ce0ea31aa/test_runner/regress/test_duplicate_layers.py#L15) that demonstrates the problem using a failpoint

However, the test in the test suite doesn't use the failpoint to induce a crash that could legitimately happen in production.
What is does instead is to return early with an `Ok()`, so that the code in Pageserver that handles duplicate layers (item 1) actually gets exercised.

That "return early" would be a bug in the routine if it happened in production.
So, the tests in the test suite are tests for their own sake, but don't serve to actually regress-test any production behavior.

# Problem 2

Further, if production code _did_ (it nowawdays doesn't!) create a duplicate layer, the code in Pageserver that handles the condition (item 1 above) is too little and too late:

* the code handles it by discarding the newer `struct Layer`; that's good.
* however, on disk, we have already overwritten the old with the new layer file
* the fact that we do it atomically doesn't matter because ...
* if the new layer file is not bit-identical, then we have a cache coherency problem
  * PS PageCache block cache: caches old bit battern
  * blob_io offsets stored in variables, based on pre-overwrite bit pattern / offsets
    * => reading based on these offsets from the new file might yield different data than before
 
# Soution

- Remove the test suite code
- Remove the Pageserver code that handles duplicate layers too late
- Add a panic/abort in the Pageserver code for when we'd overwrite a layer
  - Use `RENAME_NOREPLACE` to detect this correctly. Introduce an abstraction for that.

skyzh commented:
I once made it into a panic and then it did panic in prod 🤣 having generation numbers might help reduce the chances of such things happening but I don't feel confident enough to bring this panic back again unless we can assign each layer file a unique name. https://neondb.slack.com/archives/C04PSBP2SAF/p1689153285026749

koivunej commented on libs/utils/src/fs_ext/rename_noreplace.rs:
```diff
@@ -0,0 +1,112 @@
+        }
+        #[cfg(not(any(target_os = "linux", target_os = "macos")))]
+        {
+            std::compile_error!("OS does not support no-replace renames");
```
could hardlink + unlink?

problame replied:
can do this in future

koivunej commented on test_runner/regress/test_duplicate_layers.py:
```diff
@@ -1,143 +0,0 @@
-    env.pageserver.assert_log_contains("compact-level0-phase1-return-same")
-
-
-def test_actually_duplicated_l1(neon_env_builder: NeonEnvBuilder, pg_bin: PgBin):
```
I was thinking this should be kept around, but it just shows the absence of the problem. I guess as many of our regress tests do.

problame replied:
will re-add under different name

koivunej commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -4465,43 +4465,29 @@ impl Timeline {
-                insert_layers.push(l.clone());
+                // we should abort so we don't  otherwise we might leave some
+                error!("compaction generates a L0 layer file as output, which will cause infinite compaction.");
+                std::process::abort(); // if we returned an error, we'd not check the remaining `new_delta` layers for duplicates
```
This is not needed either. It would surely be a restart loop every 20s.

problame replied:
discussed offline

koivunej reviewed (changes requested):
I think we should keep:
- the second test case
- the noreplace
- aborting on L0 -> L1 compaction returning L0

I think we should remove:
- aborting on the situation of the first case
  - noreplace removes the dangeriousness

I think we should add:
- removal of the tempfile for which the rename failed
  - we should also remove the other files but that is non-trivial

I think we should find and reopen my issue where I described we should not make the tempfiles into `Layer` before L0 -> L1 all are ready. But that intertwines with the "is layer duplicate" check, so it's not that trivial.

But the link+unlink or rename_noreplace is a good change.

I think an easy way to hit the problem shown by the second test case is disk full; we shouldn't abort on it.

problame commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -4465,43 +4465,29 @@ impl Timeline {
+                error!("duplicate L1 layer; \
+                 likely we already overwrote the exiting file on disk with a possibly different bit pattern, \
+                 causing incoherency with in-memory copies (PS page cache) {l}");
+                std::process::abort(); // scary to run `struct Layer` desctructor (?)
```
discussed offline

problame commented:
> I think we should add:
> removal of the tempfile for which the rename failed
> we should also remove the other files but that is non-trivial

We already had/have that for deltas:

https://github.com/neondatabase/neon/blob/6ff74295b5b21d54192d20d114d78621d8d53ba0/pageserver/src/tenant/storage_layer/delta_layer.rs#L656-L665

But not for images.

(Check callers of `finish_creating()`).

I think we can move that handling code from `delta_layer.rs` to `layer.rs` so that both delta and image benefit. WDYT?

problame commented:
Pushed a commit that does it.

@koivunej I think this addresses all the concerns you raised.

Regarding `abort()` yes or no, I don't feel too strongly about it. In general, duplicates in the layer map are probably fine, i.e., lookup would do the right thing if they're present.
It's just that right now,
1. it can't be modelled safely because until recently with the `-v1-...` suffix, there was a bijection `descriptor <==> file name`, so, the file system would not allow the duplicates and the layer map was used as a check
1. legacy compaction isn't supposed to produce them, and hence
2. we probably don't have good coverage of layer map with duplicates / overlapping keys.

koivunej commented on pageserver/src/tenant/storage_layer/delta_layer.rs:
```diff
@@ -651,19 +668,11 @@ impl DeltaLayerWriter {
         timeline: &Arc<Timeline>,
         ctx: &RequestContext,
     ) -> anyhow::Result<ResidentLayer> {
-        let inner = self.inner.take().unwrap();
-        let temp_path = inner.path.clone();
-        let result = inner.finish(key_end, timeline, ctx).await;
-        // The delta layer files can sometimes be really large. Clean them up.
-        if result.is_err() {
-            tracing::warn!(
-                "Cleaning up temporary delta file {temp_path} after error during writing"
-            );
-            if let Err(e) = std::fs::remove_file(&temp_path) {
-                tracing::warn!("Error cleaning up temporary delta layer file {temp_path}: {e:?}")
-            }
-        }
-        result
```
So we wouldn't even remove the tempfile which we failed to move... Was this supposed to have a companion change...? I'll edit this soon with your response from the main PR.

problame replied:
Yeah, please resolve?

koivunej commented on libs/utils/src/fs_ext/rename_noreplace.rs:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    #[test]
+    fn test_absolute_paths() {
```
These don't need to be named like in when junit was first incepted
```suggestion
    fn absolute_paths() {
```

it would otherwise read `#[test] fn test_whatever`, no point in repeating the `test`.

koivunej commented on libs/utils/src/fs_ext/rename_noreplace.rs:
```diff
@@ -0,0 +1,112 @@
+        fs::write(&dst, b"").unwrap();
+
+        let result = rename_noreplace(&src, &dst);
+        assert!(result.is_err());
```
```suggestion
```

koivunej commented on libs/utils/src/fs_ext/rename_noreplace.rs:
```diff
@@ -0,0 +1,112 @@
+
+        fs::write(&src, b"content").unwrap();
+
+        let result = rename_noreplace(src.as_std_path(), dst.as_std_path());
+        assert!(result.is_ok());
```
```suggestion
        rename_noreplace(src.as_std_path(), dst.as_std_path()).unwrap();
```

Similarly here: `assert(res.is_ok())` produces `assertion failed` where as `unwrap` tells you what was the `Err` value.

koivunej commented on libs/utils/src/fs_ext/rename_noreplace.rs:
```diff
@@ -0,0 +1,112 @@
+    use super::*;
+
+    fn testdir() -> camino_tempfile::Utf8TempDir {
+        match crate::env::var("NEON_UTILS_RENAME_NOREPLACE_TESTDIR") {
```
Is there a reason why we don't just want to use the cargo directory for this? Looking it up, it's supposed to be used for tempfiles in test..

koivunej replied:
Hmm ok, I thought there was one but didn't have time look for it.

problame reviewed (commented):
Addressed review comments.
We don't have strict convention regarding `test_` prefix in front of tests.
I personally prefer the test_ prefix due to how "show callers" works with rust-analyzer + VScode.
If you care deeply about it, let's have a `docs/` PR or RFC, discuss it with the team, and unify the entire code base in one go. (I would of course oppose the removal of the `test_` prefix).

jcsp commented on pageserver/src/tenant/storage_layer/image_layer.rs:
```diff
@@ -917,26 +917,57 @@ impl Drop for ImageLayerWriter {
+        let tenant_conf = TenantConf {
+            gc_period: Duration::ZERO,
+            compaction_period: Duration::ZERO,
+            ..TenantConf::default()
```
TIL about this `..` syntax.  Handy!

jcsp reviewed (approved):
+1 to the rewrite test changes

problame commented:
I decided to revert the changes that made the code abort: https://github.com/neondatabase/neon/pull/7799/commits/b48c21a4aee9fdceb15f46f1702bfd009697b541

This PR achieves its goal without doing the aborts.

Ready to merge IMO
